src/simulation/elements/LSNS.cpp

```diff
@@ -52,36 +52,104 @@ int Element_LSNS::update(UPDATE_FUNC_ARGS)
-			for (ry=-2; ry<3; ry++)
+		for (rx = -2; rx < 3; rx++)
+			for (ry = -2; ry < 3; ry++)
 				if (BOUNDS_CHECK && (rx || ry))
 				{
-					r = pmap[y+ry][x+rx];
+					r = pmap[y + ry][x + rx];
 					if (!r)
 						continue;
 					rt = TYP(r);
-					if (sim->parts_avg(i,ID(r),PT_INSL) != PT_INSL)
+					if (sim->parts_avg(i, ID(r), PT_INSL) != PT_INSL)
 					{
-						if ((sim->elements[rt].Properties&PROP_CONDUCTS) && !(rt==PT_WATR||rt==PT_SLTW||rt==PT_NTCT||rt==PT_PTCT||rt==PT_INWR) && parts[ID(r)].life==0)
+						if ((sim->elements[rt].Properties&PROP_CONDUCTS) && !(rt == PT_WATR || rt == PT_SLTW || rt == PT_NTCT || rt == PT_PTCT || rt == PT_INWR) && parts[ID(r)].life == 0)
 						{
 							parts[ID(r)].life = 4;
 							parts[ID(r)].ctype = rt;
-							sim->part_change_type(ID(r),x+rx,y+ry,PT_SPRK);
+							sim->part_change_type(ID(r), x + rx, y + ry, PT_SPRK);
 						}
 					}
 				}
 	}
-	for (rx=-rd; rx<rd+1; rx++)
-		for (ry=-rd; ry<rd+1; ry++)
-			if (x+rx>=0 && y+ry>=0 && x+rx<XRES && y+ry<YRES && (rx || ry))
+	for (rx = -rd; rx < rd + 1; rx++)
+		for (ry = -rd; ry < rd + 1; ry++)
+			if (x + rx >= 0 && y + ry >= 0 && x + rx < XRES && y + ry < YRES && (rx || ry))
 			{
-				r = pmap[y+ry][x+rx];
-				if(!r)
-					r = sim->photons[y+ry][x+rx];
-				if(!r)
+				r = pmap[y + ry][x + rx];
+				if (!r)
+					r = sim->photons[y + ry][x + rx];
+				if (!r)
 					continue;
-				if (parts[ID(r)].life > parts[i].temp-273.15)
+				if (parts[ID(r)].life > parts[i].temp - 273.15)
 					parts[i].life = 1;
+			}                                                                                   //.Life serialization.(.tmp2 = turns serialization mode on)
+	bool setFilt = false;
+	int photonWl = 0;
+	for (int rx = -rd; rx <= rd; rx++)
+		for (int ry = -rd; ry <= rd; ry++)
+			if (x + rx >= 0 && y + ry >= 0 && x + rx < XRES && y + ry < YRES && (rx || ry))
+			{
+				int r = pmap[y + ry][x + rx];
+				if (!r)
+					r = sim->photons[y + ry][x + rx];
+				if (!r)
+					continue;
+				if (parts[i].tmp == 1 && TYP(r) != PT_LSNS && TYP(r) != PT_FILT)
+				{
+					setFilt = true;
+					photonWl = parts[ID(r)].life;
+				}
+			}
+	if (setFilt)
+	{
+		int nx, ny;
+		for (int rx = -1; rx <= 1; rx++)
+			for (int ry = -1; ry <= 1; ry++)
+				if (BOUNDS_CHECK && (rx || ry))
+				{
+					int r = pmap[y + ry][x + rx];
+					if (!r)
+						continue;
+					nx = x + rx;
+					ny = y + ry;
+					while (TYP(r) == PT_FILT)
+					{
+						parts[ID(r)].ctype = 0x10000000 + photonWl;
+						nx += rx;
+						ny += ry;
+						if (nx < 0 || ny < 0 || nx >= XRES || ny >= YRES)
+							break;
+						r = pmap[ny][nx];
+					}
+				}
+	}
+	// .Life Deserialization (.tmp2 = turns deserialization mode on).Sets nearby particle's life.
+	bool deserializelife = parts[i].tmp == 2;
+	for (rx = -2; rx < 3; rx++)
+		for (ry = -2; ry < 3; ry++)
+			if (BOUNDS_CHECK && (rx || ry))
+			{
+				r = pmap[y + ry][x + rx];
+				if (!r)
+					continue;
+
+				if (deserializelife && TYP(r) == PT_FILT)
```

Comment:
This if statement is wrong, so is the next one. This one sets the .life of the LSNS, which isn't supposed to happen (that would cause the LSNS to generate a spark). It should set the .life of the non-FILT particle using this ctype calculation.

In addition, I think it would be a good idea to restrict it to newlife > 0. I don't want people setting negative life. (Also, please actually use newlife instead of duplicating the ctype check twice)

Reply:
I looked at it more, I didn't fully understand the problem before so my advice was wrong. The issue I have with this is the fact that there are two if statements and LSNS is modified. LSNS shouldn't be modified, and there should only be one if statement.

How it should work is, LSNS with .tmp = 2 scans around itself for a FILT. When it finds it, it then sets the .life of nearby particles. It should either: set the .life of the particle directly opposite the LSNS; or set the .life of all particles touching it. I don't really care which. Setting the .life of the opposite particle would be easier.